src/rust/receiver-mock/src/traces/mod.rs

```diff
@@ -44,24 +45,46 @@ impl std::fmt::Display for Span {
     }
 }
 
+fn is_span_ok(span: &Span, params: &HashMap<String, String>) -> bool {
+    // TODO: This can be expanded to query by other parameters. As for now, it only filters by label.
+    for (key, value) in params.iter() {
+        if let Some(val) = span.attributes.get(key) {
+            if val.eq(value) {
+                continue;
+            }
+
+            return false;
+        } else {
+            return false;
+        }
+    }
+
+    true
+}
+
 pub fn filter_spans<'a>(spans: impl Iterator<Item = &'a Span>, params: HashMap<String, String>) -> Vec<&'a Span> {
-    spans
-        .filter(|span| {
-            // TODO: This can be expanded to query by other parameters. As for now, it only filters by label.
-            for (key, value) in params.iter() {
-                if let Some(val) = span.attributes.get(key) {
-                    if val.eq(value) {
-                        continue;
-                    }
+    spans.filter(|span| is_span_ok(span, &params)).collect()
+}
 
-                    return false;
+pub fn filter_traces<'a>(
+    traces: impl Iterator<Item = &'a Trace>,
+    spans: &'a HashMap<SpanId, Span>,
+    params: HashMap<String, String>,
+) -> Vec<Vec<&'a Span>> {
+    traces
+        .map(|trace| {
+            // Doing this functionally would be a mess if we want to handle bugs without panicking.
```

Comment:
can you elaborate?

Reply:
I wanted to do something like this:
```
trace.span_ids.map(|id| {
match spans.get(id) {
// if Some(span), then return span
// if None, print a warning and add an empty span
}
})
```
However, because we're dealing with references here (see the return type of this function), returning a newly created, empty span is not possible. To do this, we would have to pass it to this function with lifetime `'a` or longer, which isn't great. We could also resign from using references and just copy the spans, but it's not worth it in my opinion.

Reply:
On second thought, this error checking only asserts that there was no internal error, because the span will be absent in the map only if there was a bug. So, we can also ignore that case and simply do an `unwrap`.